### PR TITLE
[MESH-527] Ticker transfer based on generic authorization 

### PR DESCRIPTION
**Pull request by maxsam4**
I probably should have split this in two PRs with generic stuff in one and ticker transfer specific in the other but here we are :)

### CODE REVIEW

File: runtime/src/asset.rs [hunk 1]
```diff
@@ -244,16 +237,16 @@ decl_module! {
...
+        /// This function is used to accept a ticker transfer
+        /// NB: To reject the transfer, call remove auth function in identity module.
         ///
         /// # Arguments
         /// * `origin` It consist the signing key of the caller (i.e who signed the transaction to execute this function)
-        /// * `to_did` DID of the future owner of the ticker
-        /// * `_ticker` ticker to transfer
-        pub fn approve_ticker_transfer(origin, to_did: IdentityId, _ticker: Vec<u8>) -> Result {
+        /// * `auth_id` Authorization ID of ticker transfer authorization
+        pub fn accept_ticker_transfer(origin, auth_id: u64) -> Result {
```
Comment (maxsam4) [inline runtime/src/asset.rs]: Please just open the file and review this function as a whole. The diff is very confusing. This is essentially a new function that replaces a few of the existing functions.

File: polymesh_schema.json [hunk 2]
```diff
@@ -204,6 +204,19 @@
...
     },
+    "Authorization": {
+        "authorization_data": "AuthorizationData",
+        "authorized_by": "IdentityId",
+        "expiry": "Option<Moment>",
+        "next_authorization": "u64",
+        "previous_authorization": "u64"
+    },
+    "AuthorizationData": {
+        "_enum": {
```
Comment (vkomenda) [inline polymesh_schema.json]: I'm not sure if that makes difference but `_enum` down at line 221 uses a list contained in `[ ]` rather than in `{ }`. Which of those two is the correct notation? Or is it both?

File: primitives/src/authorization.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+/// Authorization data for two step prcoesses.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, Debug, PartialOrd, Ord)]
+pub enum AuthorizationData {
+    /// Authorization to transfer a ticker
+    TransferTicker(Vec<u8>),
+    /// Any other authorization
+    Custom(Vec<u8>),
+    /// No authorization data
+    None,
```
Comment (vkomenda) [inline primitives/src/authorization.rs]: I'm curious as for what was the reason not to use an `Option` type for this.

File: polymesh_schema.json (hunk 2 above)
Comment (maxsam4) [reply polymesh_schema.json]: Unfortunately, none of them is "correct". These are hacks to make polkadot.js work. With [] approach, enums that have arguments, don't work. with {} approach, enums that don't have arguments don't work.

Eventually, we'll not even need this file. It's a planned feature in polkadot.js to be able to fetch types directly from the node.

File: primitives/src/authorization.rs (hunk 3 above)
Comment (maxsam4) [reply primitives/src/authorization.rs]: I do not expect to use None ever in the actual world. In fact, I did not even define it in the schema. However, I added it because I needed something to be the default type and we might need none for some unique case.

Using an `Option` will add 1 bit of extra data to all the encoded values and gain us nothing.

File: primitives/src/authorization.rs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+impl Default for AuthorizationData {
```
Comment (vkomenda) [inline primitives/src/authorization.rs]: This instance would have been trivial if `Option` were used.

File: primitives/src/authorization.rs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    /// Enum that contains authorization type and data
+    pub authorization_data: AuthorizationData,
+
+    /// Identity of the organization/individual that added this authorization
+    pub authorized_by: IdentityId,
+
+    /// time when this authorization expires. optional.
+    pub expiry: Option<U>,
+
+    // Extra data to allow iterating over the authorizations.
```
Comment (vkomenda) [inline primitives/src/authorization.rs]: Missing third `/`.

File: primitives/src/authorization.rs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    /// Authorization number of the next Authorization
```
Comment (vkomenda) [inline primitives/src/authorization.rs]: Doc comments will compile into a single line without a line break, so a `.` would look better at the end of sentences.

File: primitives/src/authorization.rs (hunk 4 above)
Comment (maxsam4) [reply primitives/src/authorization.rs]: True but it doesn't really gain us anything other than saving a few lines of code. `Option` is cool but it increases the size of the variable. It also forces us to unwrap data whenever we want to use it. Thus, adding to both storage and execution costs.

Unless the "default" value represents something other than "none", I avoid using Option. For example, you'll see below that I'm using 1-indexed auth_id/auth_number rather than 0-index below because I am representing "none" as 0. I could have used Option<u64> and then used 0-index auth_id but I would have gained nothing from it except maybe a little more code clarity.

At the same time, I am using Option<expiry> because 0 can mean something other than none when it comes to timestamps. I don't think anyone will use 0 timestamps in real life but they can to show something historic that isn't valid anymore.

Basically, If I can control the data (enum/ids), I prefer to not use option and just use one specific data to represent none. However, if the data is user input, I prefer to use Option because in that case, I can not be sure that the user means 0 or none when they enter 0.

File: primitives/src/authorization.rs (hunk 5 above)
Comment (maxsam4) [reply primitives/src/authorization.rs]: This was intentional. It's an overarching code comment that applies to the next two elements. Doc comments only apply to the next element. It will not get mentioned when we generate cargo docs but I think that's fine. It's more of a code comment.

File: primitives/src/authorization.rs (hunk 6 above)
Comment (maxsam4) [reply primitives/src/authorization.rs]: Makes sense, I'll add them.

File: runtime/src/identity.rs [hunk 7]
```diff
@@ -555,6 +561,53 @@ decl_module! {
...
+            Ok(())
+        }
+
+        /// Removes an authorization
+        pub fn remove_authorization(
+            origin,
+            target_did: IdentityId,
+            auth_id: u64
+        ) -> Result {
+            let sender_key = Key::try_from(ensure_signed(origin)?.encode())?;
```
Comment (satyamakgec) [inline runtime/src/identity.rs]: Duplication of code can be avoided if you can make a public helper function that can also be called by the other modules to validate this.

File: runtime/src/asset.rs (hunk 1 above)
Comment (satyamakgec) [inline runtime/src/asset.rs]: Also please add this function in the list of dispatchable functions in the module comments

File: runtime/src/identity.rs (hunk 7 above)
Comment (maxsam4) [reply runtime/src/identity.rs]: Yeah, there's already a code hygiene task around this. Currently, I'm not even sure if this is the best way to do it. If we decide to force identities on multi_sigs, this can be simplified.

File: runtime/src/asset.rs (hunk 1 above)
Comment (maxsam4) [reply runtime/src/asset.rs]: Will do.

File: primitives/src/authorization.rs (hunk 4 above)
Comment (vkomenda) [reply primitives/src/authorization.rs]: I see your point. I would personally prefer code clarity because this all fits into a common use case for option types.

> Unless the "default" value represents something other than "none", I avoid using Option.

I misread this at first and thought you were referring to manual `Default` instances. If you instead mean `0` vs `None` then I agree: `Timestamp(0)` is not the same as `None: Option<Timestamp>`.

File: primitives/src/authorization.rs (hunk 3 above)
Comment (vkomenda) [reply primitives/src/authorization.rs]: If `None` is the member of the enum, you can assume that somebody, maybe not you, will eventually use it. I'd say the [Chekhov's gun](https://en.wikipedia.org/wiki/Chekhov%27s_gun) principle is fully applicable here. ;-)

`Option::None` is encoded as the byte `0`, and the `Option::Some` as the byte `1`. This sounds like a reasonable tradeoff which is amortised against the size of the vectors.

File: runtime/src/identity.rs [hunk 8]
```diff
@@ -778,10 +839,115 @@ decl_event!(
...
+        Self::deposit_event(RawEvent::AuthorizationRemoved(auth_id, target_did));
+    }
+
+    /// Consumes an authorization.
+    /// Checks if the auth has not expired and the caller is authorized to consume this auth.
+    pub fn consume_auth(
+        from_did: IdentityId,
+        target_did: IdentityId,
+        auth_id: u64,
+    ) -> AuthorizationStatus {
```
Comment (fmiguelgarcia) [inline runtime/src/identity.rs]: Suggestion: Replace returned type by `Result<(), AuthorizationError>`.
In that way, it follows the common design in Rust, and it will facilitate the error process. i.e:
```
match consume_auth( from, target, auth_id) {
   Ok(_) => { ... },
   Err(error) =>  print( "{:?}", error)
}
```
or 
```
if let Err(error) = consume_auth( from, target, auth_id) {
   print( "{:?}", error);
   // Or early return
}
```
And impl `Display` for `AuthorizationError` (just one place).

File: runtime/src/identity.rs [hunk 9]
```diff
@@ -555,6 +561,59 @@ decl_module! {
...
+        // Manage generic authorizations
+        /// Adds an authorization
+        pub fn add_authorization(
+            origin,
+            target_did: IdentityId,
+            authorization_data: AuthorizationData,
+            expiry: Option<T::Moment>
+        ) -> Result {
+            let sender_key = Key::try_from(ensure_signed(origin)?.encode())?;
+            let from_did =  match Self::current_did() {
```
[the remark this answers is not in the document]
Comment (maxsam4) [reply runtime/src/identity.rs]: There's a Jira task for this already :)

File: primitives/src/authorization.rs [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    /// Identity of the organization/individual that added this authorization
+    pub authorized_by: IdentityId,
+
+    /// time when this authorization expires. optional.
+    pub expiry: Option<U>,
+
+    // Extra data to allow iterating over the authorizations.
+    /// Authorization number of the next Authorization.
+    /// Authorization number starts with 1.
+    pub next_authorization: u64,
```
[the remark this answers is not in the document]
Comment (maxsam4) [reply primitives/src/authorization.rs]: Yep, but it will increase both storage and execution costs.

I don't see the need for it. See my comments above in response to Vlad's comments.

File: runtime/src/identity.rs [hunk 11]
```diff
@@ -778,10 +839,115 @@ decl_event!(
...
 
 impl<T: Trait> Module<T> {
+    pub fn add_auth(
+        from_did: IdentityId,
+        target_did: IdentityId,
+        authorization_data: AuthorizationData,
+        expiry: Option<T::Moment>,
+    ) {
+        let new_nonce = Self::multi_purpose_nonce() + 1u64;
+        // Even if this transaction fails, nonce should be increased for added unpredictability
```
Comment (adamdossa) [inline runtime/src/identity.rs]: can the transaction actually fail after this?

File: runtime/src/identity.rs [hunk 12]
```diff
@@ -778,10 +839,115 @@ decl_event!(
...
+        auth_id: u64,
+    ) -> AuthorizationStatus {
+        if !<Authorizations<T>>::exists((target_did, auth_id)) {
+            // Auth does not exist
+            return AuthorizationStatus::Invalid;
+        }
+        let auth = Self::authorizations((target_did, auth_id));
+        if auth.authorized_by != from_did {
+            // Not authorized to revoke this authorization
+            return AuthorizationStatus::Unauthorized;
```
Comment (adamdossa) [inline runtime/src/identity.rs]: Isn't this really just another case of `AuthorizationStatus::Invalid` in the sense that the authorisation you've queried for doesn't exist (i.e. there the target_did has an authorisation with the right `auth_id` but it doesn't match the `from_did`)? (realise this is nit-picking).

File: runtime/src/asset.rs [hunk 13]
```diff
@@ -244,78 +237,13 @@ decl_module! {
...
-        }
-
-        /// This function is used to transfer a ticker to someone else
-        ///
-        /// # Arguments
-        /// * `origin` It consist the signing key of the caller (i.e who signed the transaction to execute this function)
-        /// * `_ticker` ticker to transfer
-        pub fn process_ticker_transfer(origin, _ticker: Vec<u8>) -> Result {
+        /// * `auth_id` Authorization ID of ticker transfer authorization
+        pub fn accept_ticker_transfer(origin, auth_id: u64) -> Result {
```
Comment (adamdossa) [inline runtime/src/asset.rs]: Would it make sense to have a function (maybe in identity.rs) which is:
`accept_authorisation(origin, auth_id)`
which then looks at the associated AuthorisationData enum and routes to the appropriate underlying function with the right business logic to execute it?

Comment (adamdossa) [reply runtime/src/asset.rs]: This could then be batched, where we take a Vec<auth_id> and execute them as a batch.

File: runtime/src/identity.rs (hunk 11 above)
Comment (maxsam4) [reply runtime/src/identity.rs]: Nope, I'll remove the comment.

File: runtime/src/identity.rs (hunk 12 above)
Comment (maxsam4) [reply runtime/src/identity.rs]: On the business level, yes. A similar argument can be made for auth expiry as well. It's just an invalid auth if it's expired. It's a different case on the technical level though. I'd prefer to keep different return values to make sure we cover all the cases in tests.

The UI can group these errors together to show a general error if they prefer it that way.

File: runtime/src/asset.rs [hunk 14]
```diff
@@ -329,84 +257,33 @@ decl_module! {
...
-                Some(x) => x,
-                None => {
-                    if let Some(did) = <identity::Module<T>>::get_identity(&sender_key) {
-                        did
-                    } else {
-                        return Err("did not found");
-                    }
-                }
+            let ticker = match auth.authorization_data {
+                AuthorizationData::TransferTicker(_ticker) => utils::bytes_to_upper(_ticker.as_slice()),
```
[the remark this answers is not in the document]
Comment (maxsam4) [reply runtime/src/asset.rs]: Added
